Automattic/newspack-blocks
# feat(donate): checkout modal for woocommerce

### All Submissions:






### Changes proposed in this Pull Request:

This PR implements a simplified modal WooCommerce checkout for the Donate block.

When authenticated, the billing fields are removed in order to provide a more straight-forward experience:

https://user-images.githubusercontent.com/820752/215204878-e472fa09-dc47-4f61-a277-d2f43f721df3.mov

While unauthenticated, the checkout form includes ~the fields "First Name", "Last Name" and "Email"~ all fields by default. These values are defined here and are filterable so they can be customized by a wizard setting in Newspack core (Reader Revenue wizard?) or any other plugin:

https://github.com/Automattic/newspack-blocks/blob/76612d915b817efa21837ab376215138e761da8e/src/blocks/donate/frontend/class-newspack-blocks-donate-renderer.php#L36-L53

The idea is that it can still be a full checkout form, with standard address fields and the expected friction that it entails, but preserving the navigation.

At the end of the transaction, if a new reader account is created and RAS is active, the navigation is updated so the page knows it's now an authenticated reader (without the need for a page refresh).

https://user-images.githubusercontent.com/820752/215205326-e69c869a-2f45-43c7-83bf-99150b8368c6.mov

#### Update

https://github.com/Automattic/newspack-plugin/pull/2256/commits/b5c1d4936cd0e90c47211bbeeb8f2eee8bd2d97d introduced customizable billing fields:

<img width="1224" alt="image" src="https://user-images.githubusercontent.com/820752/216107697-b850cc96-7559-4efb-9293-18098a6b73bc.png">

The available fields are retrieved from `WC_Checkout()->get_checkout_fields()` so any new field should be integrated with WooCommerce in order to become available. This also allows for it to be compatible with other plugins that may add new fields.

The "Email" field is the only one not editable because it's required.

### How to test the changes in this Pull Request:

1. Checkout this branch and https://github.com/Automattic/newspack-plugin/pull/2256
2. Set the Reader Revenue platform to WooCommerce with a valid payment gateway
3. Make sure you have AMP Plus (AMP plugin on with `define( 'NEWSPACK_AMP_PLUS_ENABLED', true );`)
4. Place a "Frequency" Donate block and a "Tiers" Donate block on a page
5. In a fresh session (logged out) select an option and confirm you are able to donate as shown in the videos above
6. If you have RAS enabled, confirm the "Sign In" header link changed to "My Account"
7. Click on the page and confirm you see your order
8. Back to the page with the Donate blocks, select to donate again, and confirm the billing fields are no longer showing, like in the video above
9. Switch AMP Plus off to use vanilla AMP, click to donate, and confirm you're redirected to the regular checkout

### Other information:
